packages/lib/zuauth/src/server.ts

```diff
@@ -31,14 +66,27 @@ export async function authenticate(
   const pcd = await ZKEdDSAEventTicketPCDPackage.deserialize(serializedPCD.pcd);
 
   if (!(await ZKEdDSAEventTicketPCDPackage.verify(pcd))) {
-    throw new Error("ZK ticket PCD is not valid");
+    throw new ZuAuthAuthenticationError("ZK ticket PCD is not valid");
   }
 
   if (pcd.claim.watermark.toString() !== watermark) {
-    throw new Error("PCD watermark doesn't match");
+    throw new ZuAuthAuthenticationError("PCD watermark does not match");
+  }
+
+  if (pcd.claim.externalNullifier?.toString() !== externalNullifier) {
+    throw new ZuAuthAuthenticationError("External nullfier does not match");
+  }
+
+  // For each of the fields configured to be revealed, check that the claim
+  // contains values.
+  for (const [revealedField, fieldName] of Object.entries(revealedFields)) {
+    if (fieldsToReveal[revealedField as keyof EdDSATicketFieldsToReveal]) {
+      checkIsDefined(pcd.claim.partialTicket[fieldName], fieldName);
+    }
   }
 
   const publicKeys = config.map((em) => em.publicKey);
```

Comment:
I don't understand the problem here. Can you give me an example?

Reply:
My config says that valid tickets to look like: (pubkeyA, eventA, productA1), (pubkeyA, eventA, productA2), (pubkeyB, eventB, productB1).

I'm presented with a ticket containing (pubkeyB, eventA, productA1).  This is not actually a valid combination because the signer for B shouldn't be signing tickets from A.  However the code here accepts it.

Alternately since there's usually only one signer, I could see a ticket (pubkey, eventB, productA1).  This is invalid because it's mixing products from A with event B.  The code here accepts it anyway.

It's an unlikely issue given there's currently only one issuer (Zupass/PODBox) and it uses unique IDs.  But in theory if the signers are independent we shouldn't trust them not to step on each other.  The code here could avoid the risk by forming a set of triples of (pubkey, eventID, productID) and checking that the presented IDs match exactly one triple.  It's a bit more complicated because of the fact that some of the fields might not be revealed, so it's not a simple set membership check.

With GPC proofs, it'll be possible to prove that the triple is in a list in a ZK way directly.